contracts/container/contract.go

```diff
@@ -1298,6 +1307,139 @@ func NewEpoch(epochNum int) {
 	cleanupContainers(ctx, epochNum)
 }
 
+// PutContainerQuota sets size quota that limits all space used for storing
+// objects in cID (including object replicas). Non-positive size sets no
+// limitation. if hard is true, storage nodes decline any exceeding object at
+// PUT operation; if false, nodes are instructed to warn only, without denial
+// of service. Call must be signed by cID's owner. Limit can be changed with
+// repeated call.
+// Panics if cID is incorrect, container does not exist or new limitation already
+// exceeded (according to registered NodeReports).
+func PutContainerQuota(cID interop.Hash256, size int, hard bool) {
+	if len(cID) != interop.Hash256Len {
+		panic("wrong container id")
+	}
+	ctx := storage.GetContext()
+	ownerID := getOwnerByID(ctx, cID)
+	if ownerID == nil {
+		panic(cst.NotFoundError)
+	}
+
+	ownerSH := common.WalletToScriptHash(ownerID)
+	common.CheckOwnerWitness(ownerSH)
+
+	netmapContractAddr := storage.Get(ctx, netmapContractKey).(interop.Hash160)
+	currEpoch := contract.Call(netmapContractAddr, "epoch", contract.ReadOnly).(int)
+
+	if size > 0 {
+		var currentSize int
+		it := IterateReports(currEpoch, cID)
+		for iterator.Next(it) {
+			r := iterator.Value(it).(NodeReport)
+			currentSize += r.ContainerSize
+		}
+		if currentSize >= size {
+			panic(cst.ErrorQuotaLimitAlreadyExceeded)
+		}
+	}
+
+	key := []byte{containerQuotaKeyPrefix}
+	key = append(key, cID...)
+
+	storage.Put(ctx, key, std.Serialize(Quota{Size: size, Hard: hard}))
+	runtime.Notify("ContainerQuotaSet", cID, size, hard)
+}
+
+// ContainerQuota returns container size quota set by cID's owner. If no
+// limitation has been set, returns empty value (limitation is set to 0;
+// hard is false). Non-positive limitation must be treated as no limits.
+// Panics if cID is incorrect or container does not exist.
+func ContainerQuota(cID interop.Hash256) Quota {
+	if len(cID) != interop.Hash256Len {
+		panic("wrong container id")
+	}
+	ctx := storage.GetReadOnlyContext()
+	ownerID := getOwnerByID(ctx, cID)
+	if ownerID == nil {
+		panic(cst.NotFoundError)
+	}
+
+	key := []byte{containerQuotaKeyPrefix}
+	key = append(key, cID...)
+	v := storage.Get(ctx, key)
+	if v == nil {
+		return Quota{}
+	}
+
+	q := std.Deserialize(v.([]byte))
+
+	return q.(Quota)
+}
+
+// PutUserQuota sets size quota that limits all space used for storing objects
```

Comment:
this can also be used out of container ownership context. For example, if I have no containers, but I upload data to someone else's. Right? lookin at implementation

Reply:
tbh initially i meant that user quota == sum of every object that was put to a container this user has. i did not think about sum of objects that the user uploaded (even to other users' containers). i am not sure it can be precisely calculated how many objects a certain user has uploaded by a node, it just does not know the other node's infos. moreover,  i do not understand what user's quota means then: _this_ user set it and _this_ user uploads objects with _this_ user's wallet. what is the restriction then?

@roman-khimov

Reply:
doubt this fact. If so, to upload data, any side would need to create a container